### PR TITLE
Major Refactor: Add Ruff Linting, Nox, and Pre-Commit Hook

### PR DESCRIPTION
**Pull request by hmd101**
**Summary:**

This PR refactors the project by adding [Ruff Linting]([https://docs.astral.sh/ruff/linter)  and formatting, integrated with  [Nox](https://nox.thea.codes/en/stable/config.html) and an optional [Pre-commit](https://pre-commit.com/) hook. It enhances both code quality and development efficiency.

### Key Changes:

1. **[Ruff Linting]([https://docs.astral.sh/ruff/linter)**:
    - We have integrated [Ruff linting](https://docs.astral.sh/ruff/rules/#legend), “an extremely fast Python linter designed as a drop-in replacement for [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [pyupgrade](https://pypi.org/project/pyupgrade/), [autoflake](https://pypi.org/project/autoflake/), and more.”
2. **[Ruff Formatter](https://docs.astral.sh/ruff/formatter/):**
    - Compatible with the popular Black formatter but more performant.
3. **[Nox](https://nox.thea.codes/en/stable/)** **Automation**:
    - **Nox** has been introduced to automate the execution of our CI tasks, including Ruff linting and other testing steps. Nox provides a flexible way to manage and run our tasks in isolated virtual environments, improving the reproducibility and ease of running multiple environments and workflows.
4. **CI Pipeline Update**:
    - Our CI configuration has been updated to include Ruff linting and formatting checks. From now on, Ruff will automatically lint all new commits and pull requests to ensure they adhere to our code quality standards.
5. **Optional [Pre-commit](https://pre-commit.com/) Hook**:
    - A pre-commit hook for Ruff is now available, ensuring developers can run lint checks locally before committing code. This helps catch and resolve linting issues early in the development process.
   

### Next Steps & Open Questions:

- [x] Jupyter Notebooks: Should we include Jupyter notebooks in the Ruff linting process?  -> Yes!
- [x] Are all errors raised by the linter fixed?
- [x] Contributing file updated?

Resolves issue  #51 

------

### 1. Ruff Linting Configuration:

The current Ruff configuration includes the most commonly used linting rules. (Future PRs will likely incorporate more linting rules, as discussed [here](https://github.com/plenoptic-org/plenoptic/issues/51#issuecomment-2237167596)).

- **Pycodestyle (`E`)**: Enforces style consistency based on PEP8.
- **Pyflakes (`F`)**: Detects basic errors such as undefined variables or missing imports.
- **Pyupgrade (`UP`)**: Ensures usage of modern Python syntax, making our code more future-proof.
- **Flake8-simplify (`SIM`)**: Improves code readability by suggesting simplified expressions. We have chosen to ignore rule `SIM105` since our team prefers explicit boolean comparisons in some cases.
- **Isort (`I`)**: Ensures imports are organized and ordered consistently, improving code readability and maintainability.

Note that a couple of linting rules were **ignored** with  `#noqa <ruff rule>`  (ignoring a single line) and  `#ruff: noqa <ruff rule>` (ignoring entire file):

- `E501` for too long lines has been ignored several times to ensure for example that links will still work or flow charts remain readable
- `I001`  sorting imports has been ignored once in `src/plenoptic/__init__.py` as the import order matters to avoid circular dependencies and tests to fail
- `F401`  unused imports
    - `# ruff: noqa: F401` is used in several `__init__.py` files to ignore unused imports. When unused imports are removed, tests fail.
 - `SIM105` : [Checks](https://docs.astral.sh/ruff/rules/suppressible-exception/) for `try`-`except`-`pass` blocks that can be replaced with the `contextlib.suppress` context manager.
    - currently ignored globally in `pyproject.toml`
- `F403` flags wildcard imports `*` , which is bad practice. (This is part of the issue #246 and notes have been added there)
    
    To remove wildcard (`*`) imports, the 2 recommended approaches are:
    
    1. **Explicit Imports in the `__init__.py` File**:
        - Manually specify the functions and classes to be imported.
        - Example: `from .validate import remove_grad`
        - This approach aligns with the flat hierarchy recommendation from the OpenSci review (see issue #246)
    2. **Nested Structure with Subpackages/Modules**:
        - Organize the code into subpackages and import modules explicitly.
        - Example: `plenoptic.tools.validate`
            - Use `from . import validate` within the subpackage.
        - In files like `validate.py`, define a list called `__all__ = ["remove_grad", ...]`, which specifies what can be imported from the file.
            - This can live in either the module file itself or the `__init__.py`, and it controls the public API of that module.

### 2. [Ruff Formatting](https://docs.astral.sh/ruff/configuration/)
Currently, formatting is enforced to emulate Black in  `pyproject.toml`: 
```bash
indent-width = 4 # same as Black
[tool.ruff.format]
# Like Black, use double quotes for strings.
quote-style = "double"

# Like Black, indent with spaces, rather than tabs.
indent-style = "space"

# Like Black, respect magic trailing commas.
skip-magic-trailing-comma = false

# Like Black, automatically detect the appropriate line ending.
line-ending = "auto"

```

## Nox & Pre-commit Hooks and their Configurations:

**TL;DR:**

- Both tools are complementary and contribute to better maintainability by promoting consistent coding practices and making it easier to identify and fix issues early in the development process. **Both are recommended by the [Scientific Python Library Development Guide](https://learn.scientific-python.org/development/guides/style/).**
- **Nox**: A task-runner automating testing across different environments *after* commits, ensuring code works well in varied setups specified in `noxfile.py`.
- **[Pre-commit](https://pre-commit.com/)**: Ensures code quality *before* commits by enforcing checks like formatting and linting. For this to work, pre-commit needs to be installed with `pip install pre-commit` . Pre-commit checks can be ignored, by adding `--no-verify`  to your commit message.

### 2. Nox Configuration:

Nox is a Python-based task runner that is used to automate various development workflows, such as testing, building, and deploying applications. It allows you to define sessions, which are Python functions in the `noxfile.py` decorated with `@nox.session`  with various parameters, such as `name`, by which these sessions can then be executed via the command line. This easily allows to test code by running pre-defined tasks with a single command across multiple Python versions without manually setting up environments.

**Example:** Based on the following  `noxfile.py` , the command `nox -s tests`  would run the session named "tests” below.  `nox`  without any commands would run all the sessions defined in the `noxfile.py` .

```python
@nox.session(name="tests", python=["3.10", "3.11", "3.12"])
def tests(session):
    # run tests
    session.install("pytest")
    session.run("pytest")
```

Here are the current Nox sessions available for execution:

- `nox -s lint`
- `nox -s tests`
- `nox -s coverage`

### 3. Pre-Commit Configurations:

**Hook:**
A [pre-commit hook](https://pre-commit.com/) can be added to run Ruff linting on code *before* it is committed. This ensures that developers catch and fix linting issues early in the development process. **It only runs when [installed](https://pre-commit.com/#install) and can be bypassed with `git commit -m <my commit message> --no-verify`.**

**Considered Configurations:**  (in `.pre-commit.yaml`:
(**Legend:** ✅: included, ❓: [recommended](https://learn.scientific-python.org/development/guides/style/), but not included, possibly in future PR)

- `check-added-large-files` ✅
    - **Purpose**: Checks for large files added to the repository, typically to prevent accidental inclusion of large binaries or datasets.
- `check-case-conflict` ✅
    - **Purpose**: Detects potential filename conflicts due to case-insensitive filesystems (e.g., Windows) where `File.txt` and `file.txt` would be considered the same
- `check-merge-conflict` ✅
    - **Purpose**: Checks for leftover merge conflict markers (`<<<<<<<`, `=======`, `>>>>>>>`) in files.
- `check-symlinks` ❓
    - **Purpose**: Verifies that symbolic links are valid and do not point to non-existent targets.
- `check-yaml`✅
    - **Purpose**: Validates YAML files for syntax errors.
- `debug-statements` ✅
    - **Purpose**: Detects debug statements (e.g., `print`, `console.log`) left in code.
- `end-of-file-fixer`✅
    - **Purpose**: Ensures files have a newline at the end.
- `trailing-whitespace`✅
    - **Purpose**: Removes trailing whitespace characters from files.

### CI Pipeline:

Finally, the CI configuration `ci.yml` has been updated to include a [step](https://github.com/ChartBoost/ruff-action) each for running Ruff *formatting* and *linting* check on all new commits and pull requests. Two separate actions were chosen to signal the developer which combination of the two might be causing an error.

### CODE REVIEW

**Review by billbrod — DISMISSED** [review]

**Review by billbrod — CHANGES_REQUESTED** [review]
Thanks for working on this! We'll talk about this more when we meet, but I think:
- Add all the checks you discuss in #51 . At least try them out and, if any seem like really major changes / you're not sure, we can revisit them.
- In the ruff config, include the descriptions you wrote in the above issue for each check.
- For the CI, I'd rather not use a separate action for this. It will be installed as part of `[dev]` optional dependnecies, so we can run it directly (similar to what we do for pytest)
- Does pre-commit have a check that can be run post-hoc based on the included config? I'd like to include instructions on how to do that, and do so as part of our CI
- In the CI, the linter step needs to be added in the `needs` field for the `check` step. This will make it required for merging.
- Currently, we're excluding notebooks from the linter. I'd like to check what the linter does to them, to see if it looks reasonable, before deciding on that completely.

Then we will need to run the linter as part of this PR. My advice would be to run each of the checks, one at a time, to understand what changes are made. Then make any changes necessary and put each in a separate commit. Keep track of these commits hashes and add them to `.git-blame-ignore-revs` in the root directory (see [docs](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)), with comments.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -199,7 +257,124 @@ several choices for how to run a subset of the tests:
...
+
+`nox` offers a variety of configuration options, you can learn more about it from their
+[documentation](https://nox.thea.codes/en/stable/config.html).
+
+#### Multi-python version testing with pyenv
+Sometimes, before opening a pull-request that will trigger the `.github/workflow/ci.yml` continuous
+integration workflow, you may want to test your changes over all the supported python versions locally.
+
+Handling multiple installed python versions on the same machine can be challenging and confusing.
+[`pyenv`](https://github.com/pyenv/pyenv) is a great tool that really comes to the rescue.
```
Comment (billbrod) [inline CONTRIBUTING.md:301]:
```suggestion
[`pyenv`](https://github.com/pyenv/pyenv) is a great tool that really comes to the rescue. Note that `pyenv` just handles python versions --- virtual environments have to be handled separately, using [`pyenv-virtualenv`](https://github.com/pyenv/pyenv-virtualenv)!
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -199,7 +257,124 @@ several choices for how to run a subset of the tests:
...
+2. Run nox specifying the python version as an option.
+    ```bash
+    nox -p 3.10
+    ```
+
+Note that `noxfile.py` lists the available option as keyword arguments in a session specific manner.
+
+If you have multiple python version installed, we recommend to manage your virtual environments
+through `pyenv`. For that you'll need to install the extension
+[`pyenv-virtualenv`](https://github.com/pyenv/pyenv-virtualenv).
```
Comment (billbrod) [inline CONTRIBUTING.md:351]:
```suggestion
As mentioned earlier, if you have multiple python version installed, we recommend you manage your virtual environments through `pyenv` using the [`pyenv-virtualenv`](https://github.com/pyenv/pyenv-virtualenv) extension.
```

File: noxfile.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from pathlib import Path
+
+import nox
+
+
+@nox.session(name="lint")
+def lint(session):
+    # run linters
+    session.install("ruff")
+    session.run("ruff", "check", "--ignore", "D")
```
Comment (billbrod) [inline noxfile.py:10]: why `--ignore D`?

File: noxfile.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@nox.session(name="tests", python=["3.10", "3.11", "3.12"])
+def tests(session):
+    # run tests
+    session.install("pytest")
+    # Install pytest-cov for coverage reporting
+    session.install("pytest-cov")
+    # Install dependencies listed in pyproject.toml
+    session.install(
```
Comment (billbrod) [inline noxfile.py:20]: instead of all this install and pythonpath manipulation, do `session.install(".[dev]")` (see [here](https://learn.scientific-python.org/development/guides/tasks/#introduction))

File: pyproject.toml [hunk 5]
```diff
@@ -57,6 +57,8 @@ dev = [
     'pytest-cov',
     'pytest-xdist',
     "pooch>=1.2.0",
+    "nox",
```
Comment (billbrod) [inline pyproject.toml:60]: since `nox` is optional, I don't think it should be included here (just as pre-commit isn't listed here)

File: pyproject.toml [hunk 6]
```diff
@@ -84,3 +86,78 @@ local_scheme = 'no-local-version'
 addopts = "--cov=plenoptic -n auto"
 testpaths = ["tests"]
 
+[tool.ruff]
+extend-include = ["*.ipynb"]
+src = ["src", "tests"]
```
Comment (billbrod) [inline pyproject.toml:91]: and `examples/` right?

File: pyproject.toml [hunk 7]
```diff
@@ -84,3 +86,78 @@ local_scheme = 'no-local-version'
...
+# Exclude a variety of commonly ignored directories.
+exclude = [
```
Comment (billbrod) [inline pyproject.toml:93]: I think most of these can be removed. maybe except `.git` and `.github`, but ruff automatically ignores whatever's in `.gitignore`. for those here that are relevant (e.g., not `node_modules`), they belong in gitignore instead.

File: pyproject.toml [hunk 8]
```diff
@@ -84,3 +86,78 @@ local_scheme = 'no-local-version'
...
+select = [
+    # pycodestyle
+    "E",
+    # Pyflakes: basic static analzsis for common errors like undefined names
+    # and missing imports.
+    "F",
+    # pyupgrade
+    "UP",
+    # flake8-bugbear
+    # "B",
```
Comment (billbrod) [inline pyproject.toml:154]: remove commented out code. Is there a ruff or precommit for finding commented out code?
```suggestion
```

File: pyproject.toml [hunk 9]
```diff
@@ -84,3 +86,78 @@ local_scheme = 'no-local-version'
...
+    # flake8-simplify
+    "SIM",
+    # isort
+    "I",
+]
+ignore = ["SIM105"]
```
Comment (billbrod) [inline pyproject.toml:160]: I'm now revisiting this decision -- I don't remember why we decided on ignoring this in the first place, do you?

File: pyproject.toml [hunk 10]
```diff
@@ -84,3 +86,78 @@ local_scheme = 'no-local-version'
...
+
+[tool.ruff.lint.pydocstyle]
```
Comment (billbrod) [inline pyproject.toml:162]: unnecessary right now, since we haven't activated ruff's docstring linting

File: src/plenoptic/data/fetch.py [hunk 11]
```diff
@@ -5,56 +5,61 @@
...
+    "portilla_simoncelli_synthesize_torch_v1.12.0_ps-refactor-2.npz": OSF_TEMPLATE.format(  # noqa: E501
+        "en8du"
+    ),
 }
 DOWNLOADABLE_FILES = list(REGISTRY_URLS.keys())
 
-import pathlib
-from typing import List
+# ignore E402
+import pathlib  # noqa: E402
```
Comment (billbrod) [inline src/plenoptic/data/fetch.py:61]: I think this can be fixed by moving all the imports above the constant definitions. There's no real reason why I did it in this order.

File: src/plenoptic/metric/classes.py [hunk 12]
```diff
@@ -42,4 +44,5 @@ def forward(self, image):
         # (down-sampled by factors of 2). To combine these into one
         # vector, we need to flatten each of them and then unsqueeze so
         # it is 3d
-        return torch.cat([i.flatten() for i in activations]).unsqueeze(0).unsqueeze(0)
+
+        return torch.cat([i.flatten() for i in activations])[None, None, :]
```
Comment (billbrod) [inline src/plenoptic/metric/classes.py:48]: I don't love this syntax -- what rule is it coming from?

File: src/plenoptic/metric/perceptual_distance.py [hunk 13]
```diff
@@ -364,38 +384,47 @@ def normalized_laplacian_pyramid(img):
     (_, channel, height, width) = img.size()
 
     N_scales = 6
-    spatialpooling_filters = np.load(DIR / 'DN_filts.npy')
-    sigmas = np.load(DIR / 'DN_sigmas.npy')
+    spatialpooling_filters = np.load(os.path.join(DIRNAME, "DN_filts.npy"))
```
Comment (billbrod) [inline src/plenoptic/metric/perceptual_distance.py:387]: why this change? I'm trying to move away from os.path.join to using pathlib

File: src/plenoptic/metric/perceptual_distance.py [hunk 14]
```diff
@@ -418,30 +447,41 @@ def nlpd(img1, img2):
 
     References
     ----------
-    .. [1] Laparra, V., Ballé, J., Berardino, A. and Simoncelli, E.P., 2016. Perceptual image quality
-       assessment using a normalized Laplacian pyramid. Electronic Imaging, 2016(16), pp.1-6.
+    .. [1] Laparra, V., Ballé, J., Berardino, A. and Simoncelli, E.P., 2016. Perceptual
+    image quality assessment using a normalized Laplacian pyramid. Electronic Imaging,
```
Comment (billbrod) [inline src/plenoptic/metric/perceptual_distance.py:451]: this breaks sphinx's rendering of the reference. is it coming from one of the ruff rules?

File: src/plenoptic/simulate/canonical_computations/filters.py [hunk 15]
```diff
@@ -67,10 +65,7 @@ def circular_gaussian2d(
         Circular gaussian kernel, normalized by total pixel-sum (_not_ by 2pi*std).
         `filt` has `Size([out_channels=n_channels, in_channels=1, height, width])`.
     """
-    if isinstance(std, float):
-        device = torch.device("cpu")
-    else:
-        device = std.device
+    device = torch.device("cpu") if isinstance(std, float) else std.device
```
Comment (billbrod) [inline src/plenoptic/simulate/canonical_computations/filters.py:68]: this is preferred?

File: src/plenoptic/simulate/models/frontend.py [hunk 16]
```diff
@@ -339,10 +360,15 @@ def __init__(
     ):
         super().__init__()
         if pretrained:
-            assert kernel_size in [31, (31, 31)], "pretrained model has kernel_size (31, 31)"
+            assert kernel_size in [
```
Comment (billbrod) [inline src/plenoptic/simulate/models/frontend.py:363]: this is ugly. I should probably just rewrite it to use if so it formats nicer...

File: src/plenoptic/synthesize/mad_competition.py [hunk 17]
```diff
@@ -331,7 +366,7 @@ def _check_convergence(self, stop_criterion, stop_iters_to_check):
          Have we been synthesizing for ``stop_iters_to_check`` iterations?
          | |
         no yes
-         | '---->Is ``abs(synth.loss[-1] - synth.losses[-stop_iters_to_check]) < stop_criterion``?
+         | '---->Is abs(synth.loss[-1] - synth.loss[-stop_iters_to_check]) < stop_crit?
```
Comment (billbrod) [inline src/plenoptic/synthesize/mad_competition.py:369]: why this change?

File: src/plenoptic/synthesize/metamer.py [hunk 18]
```diff
@@ -274,7 +298,7 @@ def _check_convergence(self, stop_criterion, stop_iters_to_check):
          Have we been synthesizing for ``stop_iters_to_check`` iterations?
          | |
         no yes
-         | '---->Is ``abs(synth.loss[-1] - synth.losses[-stop_iters_to_check]) < stop_criterion``?
+         | '----> Is the change in loss < stop_criterion over ``stop_iters_to_check``?
```
Comment (billbrod) [inline src/plenoptic/synthesize/metamer.py:301]: why this change?

File: src/plenoptic/synthesize/metamer.py [hunk 19]
```diff
@@ -798,18 +878,23 @@ def _closure(self) -> Tuple[Tensor, Tensor]:
...
 
          Have we been synthesizing for ``stop_iters_to_check`` iterations?
          | |
         no yes
-         | '---->Is ``abs(self.loss[-1] - self.losses[-stop_iters_to_check] < stop_criterion``?
+         | '---->Is the change in loss < stop_criterion over ``stop_iters_to_check``?
          |      no |
          |       | yes
-         |-------' '---->Have we synthesized all scales and done so for ``ctf_iters_to_check`` iterations?
+         |-------' '---->Are all scales synthesized for `ctf_iters_to_check` iterations?
```
Comment (billbrod) [inline src/plenoptic/synthesize/metamer.py:897]: why this change?

**Review by billbrod — CHANGES_REQUESTED** [review]
Thanks for putting together this (huge) PR Hanna!

- It would be good to run pre-commit as part of the CI too -- when I pulled down the branch and ran it, it made some additional changes (not related to ruff, but the other hooks, which I think we'd also want to run regularly). From poking around, it looks like it's impossible to run something like "pre-commit check", which would check whether the hooks would pass and fail, but not change anything. Therefore, I think the only real CI option is [pre-commit.ci](https://pre-commit.ci/), which will also push back to the repo. It's not ideal (I'd like to have as much as possible explicit in the git repo, rather than involving stuff that has to be configured through separate websites), but still fine I think. Does that match your understanding as well?
- The contributing document has gotten complex as part of this. I think it needs a run through to improve the flow, but that's not part of this PR
- Can you add your notes about the wildcard imports and the two approaches to #246?
- it looks like the [right way](https://docs.astral.sh/ruff/rules/unused-import/) to handle F401 in __init__ files is with `__all__`, which we should do anyway. I added a note to #246.
- Can you move the next steps of your PR note to a relevant issue? maybe #51 ?
- you have several comments of the form
    ```python
    # ignore F401
    # ruff: noqa: F401
    ```
    or similar. is the ignore line doing anything there? I don't think it adds any info beyond the noqa line (which we do explain in contributing), so if it's not doing anything functional, it should be removed
- There are several changes that I think were not made by ruff or pre-commit? I flagged them, asking "why this change". I'm not necessarily opposed to them, but want to double-check whether they were made directly by you or by ruff and, if by you, want to understand your thought process

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -272,6 +297,25 @@ nox
...
 
+If you only want to run an individual session (e.g., lint or test), you can first check which sessions are available with the following command:
+
+```bash
+nox -l
+```
+or
+```bash
+nox -list
+```
```
Comment (billbrod) [inline CONTRIBUTING.md:308]:
```suggestion
```

**Review by billbrod — APPROVED** [review]
This looks good! I'm pushing a couple of small changes, and then I'll go ahead and merge this.

Following up from my earlier comment: we've gone ahead and added pre-commit.ci.
